zerver/lib/rate_limiter.py

```diff
@@ -307,3 +310,36 @@ def __init__(self, entity: RateLimitedObject, secs_to_freedom: float, over_limit
 
 def get_rate_limit_result_from_request(request: HttpRequest, entity_type: str) -> RateLimitResult:
     return request._ratelimit[entity_type]
+
+def rate_limit_event_by_entity(request: HttpRequest, entity: RateLimitedUser) -> None:
+    max_window = rules['api_by_user'][0][0]
+    max_calls = rules['api_by_user'][0][1]
+    user = entity.user
+
+    if user.id not in events_ttl_cache:
+        events_ttl_cache[user.id] = [max_calls, time.time()]
+
+    allowance, last_check = events_ttl_cache[user.id]
+    current = time.time()
+
+    time_passed = current - last_check
+    last_check = current
+    allowance += time_passed * (max_calls / max_window)
+
+    if (allowance > max_calls):
+        allowance = max_calls
+
+    if (allowance < 1.0):
+        entity_type = type(entity).__name__
+        if not hasattr(request, '_ratelimit'):
+            request._ratelimit = {}
+        request._ratelimit[entity_type] = RateLimitResult(
+            entity=entity,
+            secs_to_freedom=max_window-(max_window*allowance),
+            over_limit=True
+        )
+        raise RateLimited(entity_type)
+    else:
+        allowance -= 1.0
+
+    events_ttl_cache[user.id] = [allowance, last_check]
```

Comment:
This implementation needs (1) a docstring explaining what algorithm it's trying to do (I haven't tried too hard to read it, but this should have a clear explanation of the algorithm and its expected behavior), (2) automated tests, and (3) to handle setting the full rate limiter API by setting the fields needed by `RateLimitMiddleware`.

Reply:
Working on it.